tests/boards/native_posix/cpu_wait/src/main.c

```diff
@@ -0,0 +1,295 @@
+/*
+ * Copyright (c) 2020 Oticon A/S
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <ztest.h>
+#include <zephyr.h>
+#include <sys/printk.h>
+
+#include <stdlib.h>
+#include <stdio.h>
+#if defined(CONFIG_BOARD_NATIVE_POSIX)
+#include "board_soc.h"
+#endif
+
+/**
+ * @brief Basic test of the POSIX arch k_busy_wait() and cpu_hold()
+ * functions
+ *
+ * In this basic case, only one k_busy_wait() or posix_cpu_hold executes
+ * at a time
+ */
+static void test_cpu_hold_basic(void)
+{
+	uint32_t wait_times[] =  {1, 30, 0, 121, 10000};
+	uint64_t time2, time1 = posix_get_hw_cycle();
+
+	for (int i = 0; i < sizeof(wait_times)/sizeof(uint32_t); i++) {
+		k_busy_wait(wait_times[i]);
+		time2 = posix_get_hw_cycle();
+		zassert_true(time2 - time1 == wait_times[i],
+				"k_busy_wait failed %llu-%llu!=%u\n",
+				time2, time1, wait_times[i]);
+		time1 = time2;
+	}
+
+	for (int i = 0; i < sizeof(wait_times)/sizeof(uint32_t); i++) {
+		posix_cpu_hold(wait_times[i]);
+		time2 = posix_get_hw_cycle();
+		zassert_true(time2 - time1 == wait_times[i],
+				"posix_cpu_hold failed %llu-%llu!=%u\n",
+				time2, time1, wait_times[i]);
+		time1 = time2;
+	}
+}
+
+#define WASTED_TIME 1000 /* 1ms */
+#define THREAD_PRIO 0
+#define THREAD_DELAY 0
+
+#define ONE_TICK_TIME (1000000ul / CONFIG_SYS_CLOCK_TICKS_PER_SEC)
+#define ONE_AND_HALF_TICKS  (ONE_TICK_TIME + (ONE_TICK_TIME>>1))
+
+static void thread_entry(void *p1, void *p2, void *p3);
+
+K_THREAD_DEFINE(TIME_WASTER, CONFIG_ARCH_POSIX_RECOMMENDED_STACK_SIZE,
+		thread_entry, 0, 0, 0,
+		THREAD_PRIO, 0, THREAD_DELAY);
+K_SEM_DEFINE(start_sema, 0, 1);
+
+/**
+ * Thread meant to come up and waste time during the k_busy_wait() and
+ * posix_cpu_hold() calls of test_cpu_hold_with_another_thread()
+ *
+ * As the test condition is all about the respective timing,
+ * when different points in the thread are reached are left as
+ * comments below (All for the default system tick of native_posix of 10ms)
+ */
+void thread_entry(void *p1, void *p2, void *p3)
```

Comment:
The timing of this thread and the actual test thread is tightly coupled. Maybe changing to something like the pseudo-code below could make the test case more robust towards future changes?

WASTER_THREAD:
```
for 1..number_of_tests {
     k_sem_take(&start_sema, K_FOREVER);
     k_sleep(Z_TIMEOUT_TICKS(1));
     k_busy_wait(WASTED_TIME);
     k_sem_give(&end_sema);
}
```

TEST_THREAD:
```
test 1:
k_sem_give(&start_sema);
time1 = posix_get_hw_cycle();
k_busy_wait/posix_cpu_hold(...)
time2 = posix_get_hw_cycle();
assert time2-time1 == ...
k_sem_take(&end_sema, K_FOREVER);

test 2:
k_sem_give(&start_sema);
...
k_sem_take(&end_sema, K_FOREVER);

...
```

Reply:
>The timing of this thread and the actual test thread is tightly coupled. 

Yep, absolutely. 

> could make the test case more robust towards future changes?

I don't think this can be changed to be really safe for future changes. I mean, the whole thing this test is doing is to cause an interrupt (or the other thread) to come just in the right moment to cause a loss of context just before the return of k_busy_wait/posix_cpu_hold, or  in a place where it some other place where it should matter/not matter, and test that against the exact expected behavior at that point.
The only way to make this test safe to changes is to only test  with bigger or equal conditions, but doing that would mean testing very little, and letting bugs like the one I found (and fixed in this PR) slip by.

Reply:
The point of my suggested change would be to avoid having the whole timeline unrolled in the `WASTER_THREAD` and to chop up the timeline into smaller pieces by syncing the two threads between subtests. That should make it a little less error prone to add a new subtest, because the timing of one subtest would not mess with the timing of other subtests.

Reply:
I understand. But will adding the extra semaphore synchronization really help? I mean, if the timing changes, that particular test will still break and somebody will need to adjust it in the same way. With the semaphore the following tests would not break, but.. they won't get to execute anyhow. Or?

Reply:
A subtest can pass but still bring the timelines of the two threads out of sync and then causing subtests that run _after_ it to fail. That kind of dependency can make debugging quite difficult.